src/csharp/Grpc.Core/Internal/ServerCallHandler.cs

```diff
@@ -292,11 +292,20 @@ public Task HandleCall(ServerRpcNew newRpc, CompletionQueueSafeHandle cq)
 
     internal static class HandlerUtils
     {
-        public static Status StatusFromException(Exception e)
+        public static Status StatusFromException(Exception e, Metadata callContextResponseTrailers)
         {
             var rpcException = e as RpcException;
             if (rpcException != null)
             {
+                // There are two sources of metadata entries on the server-side:
+                // 1. serverCallContext.ResponseTrailers
+                // 2. trailers in RpcException thrown by user code in server side handler.
+                // As metadata allows duplicate keys, the logical thing to do is
+                // to just merge trailers from RpcException into serverCallContext.ResponseTrailers.
+                foreach (var entry in rpcException.Trailers)
```

Comment:
an optional, minor note in that case: perhaps rename to `MergeTrailersAndGetStatusFromException`